v2/controller.go

```diff
@@ -0,0 +1,212 @@
+	for _, o := range strs {
+		if err := o.write(path, defaultFilePerm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func NewController(mountpoint string, g string, resources *Resources) (*Controller, error) {
+	if err := VerifyGroupPath(g); err != nil {
+		return nil, err
+	}
+
+	path := filepath.Join(mountpoint, string(g))
+	if err := os.MkdirAll(path, defaultDirPerm); err != nil {
+		return nil, err
+	}
+	if err := writeCGStr(path, resources.Values()); err != nil {
+		// clean up cgroup dir on failure
+		os.Remove(path)
+		return nil, err
+	}
+	return &Controller{
+		unifiedMountpoint: mountpoint,
+		path:              path,
+	}, nil
+}
+
+func LoadController(mountpoint string, g string) (*Controller, error) {
+	if err := VerifyGroupPath(g); err != nil {
+		return nil, err
+	}
+	path := filepath.Join(mountpoint, string(g))
+	return &Controller{
+		unifiedMountpoint: mountpoint,
+		path:              path,
+	}, nil
+}
+
+type Controller struct {
+	unifiedMountpoint string
+	path              string
+}
+
+func (c *Controller) NewChild(name string, resources *Resources) (*Controller, error) {
+	if strings.HasPrefix(name, "/") {
+		return nil, errors.New("name must be relative")
+	}
+	path := filepath.Join(c.path, name)
+	if err := os.MkdirAll(path, defaultDirPerm); err != nil {
+		return nil, err
+	}
+	if err := writeCGStr(path, resources.Values()); err != nil {
+		// clean up cgroup dir on failure
+		os.Remove(path)
+		return nil, err
+	}
+	return &Controller{
+		unifiedMountpoint: c.unifiedMountpoint,
+		path:              path,
+	}, nil
+}
+
+func (c *Controller) AddProc(pid uint64) error {
+	v := CGStr{
+		key:   cgroupProcs,
+		value: pid,
+	}
+	return writeCGStr(c.path, []CGStr{v})
+}
+
+func (c *Controller) Delete() error {
+	return remove(c.path)
+}
+
+func (c *Controller) Procs(recursive bool) ([]uint64, error) {
+	var processes []uint64
+	err := filepath.Walk(c.path, func(p string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !recursive && info.IsDir() {
+			if p == c.path {
+				return nil
+			}
+			return filepath.SkipDir
+		}
+		_, name := filepath.Split(p)
+		if name != cgroupProcs {
+			return nil
+		}
+		procs, err := parseCgroupProcsFile(p)
+		if err != nil {
+			return err
+		}
+		processes = append(processes, procs...)
+		return nil
+	})
+	return processes, err
+}
+
+func (c *Controller) Freeze() error {
+	return c.freeze(c.path, Frozen)
+}
+
+func (c *Controller) Thaw() error {
+	return c.freeze(c.path, Thawed)
+}
+
+func (c *Controller) freeze(path string, state State) error {
+	values := state.Values()
+	for {
+		if err := writeCGStr(path, values); err != nil {
+			return err
+		}
+		current, err := fetchState(path)
+		if err != nil {
+			return err
+		}
+		if current == state {
+			return nil
+		}
+		time.Sleep(1 * time.Millisecond)
+	}
+}
+
+func (c *Controller) Stat(g string) (*statsv2.Metrics, error) {
+	var stats statsv2.Metrics
+	return &stats, nil
```

Comment:
Sorry, I didn't get it, how do you plan to collect stats for different controllers?

Reply:
cgroupv2 will list enabled controllers in `cgroup.controllers` file. I think it is easy to collect the data for the controllers.

Reply:
As for the implementation, i'm still working on a simpler way to the files and parse them easily without a lot of switches.  For now, we can just omit stats right now.